### PR TITLE
Changes in how paths are handled

### PR DESCRIPTION
**Pull request by ret2libc**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [-] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This is some preliminary work for rz-pm and plugins handling in general. Before working on rz-pm I want to have paths in rizin well defined and clear.

Changes:
* remove RIZIN_EXTRAS define because useless IMO. extras plugin should just be handled the same way as other plugins, no need to have a special "extras" thing
* do not use `<version>` in `/usr/share` but just assume one version is installed. You can install one version per prefix at a time anyway. This is similar to what is done by all other linux tools.
* plugins are now in `/usr/lib64/rizin/plugins` and `~/.local/lib64/rizin/plugins`
* for one release or so we keep supporting home plugins from `~/.local/share/rizin/plugins`, to avoid breaking things immediately
* introduce rz_path in rz_util to group together functions that work on Rizin paths.
* simplify rz_userconf.h.in to have less defines. Whether a path is home vs system specific is chosen by the code, by using `rz_path_system` vs `rz_path_home`. This way the same structure is respected in the home and in the system directories.
* rz_lib now keeps track of which directories have been loaded, to avoid trying to re-load the same things again and again (e.g. if you do a home installation the home plugins and the "system" plugins are the same. Also `dir.plugins` is the same usually). There is now no command to load plugins from a given directory, but if there is the API to load a directory already has a `force` option that can force loading of a directory already loaded anyway.

**Test plan**

- Try to load/list plugins
- Try to load/list themes

**Closing issues**

...

### CODE REVIEW

**Comment by thestr4ng3r** [comment]
> plugins are now in /usr/lib64/rizin/plugins and ~/.local/lib64/rizin/plugins

Are we sure we want this? What is the disadvantage of just `.../lib[64]/rizin`? Such a change would require updates in all external plugins, so if we do it I think we should at least have both paths supported for a certain time period.

**Comment by ret2libc** [comment]
> > plugins are now in /usr/lib64/rizin/plugins and ~/.local/lib64/rizin/plugins
> 
> Are we sure we want this? What is the disadvantage of just `.../lib[64]/rizin`? 

Please note that in current `dev` plugins are in `/usr/lib64/rizin/0.3.0` (i have never seen a plugin installed in this directory though...) or in `~/.local/share/rizin/plugins`. Having `plugins` dir or not doesn't change the fact that we break previous location, so I think it's just "clearer" that those things are plugins if we call the directory plugins :P

The change aims to:
1) Use the directories in the right way. In particular `share` should not be used for `.so` files.
2) Uniform the ways `~/.local` and `/usr` are used.

> Such a change would require updates in all external plugins, so if we do it I think we should at least have both paths supported for a certain time period.

We could support both for 1/2 releases if we want, yes. However.... how many plugins currently work with Rizin? 5? 6? If there are more than 10 I can add compatibility support, otherwise it's probably easier to just update those plugins (e.g. rz-ghidra, jsdec, rz-keystone, what else?)

**Comment by thestr4ng3r** [comment]
> Please note that in current `dev` plugins are in `/usr/lib64/rizin/0.3.0` (i have never seen a plugin installed in this directory though...) or in `~/.local/share/rizin/plugins`. Having `plugins` dir or not doesn't change the fact that we break previous location, so I think it's just "clearer" that those things are plugins if we call the directory plugins :P

Makes sense. Since now (pre 1.0) is the best time to still do such changes, I agree with the change.

> We could support both for 1/2 releases if we want, yes.

Let's do a period of 1 release imo. Just to not have a hard break in a single commit.

File: librz/analysis/arch_platform.c [hunk 1]
```diff
@@ -128,17 +128,18 @@ RZ_API bool rz_arch_load_platform_sdb(RZ_NONNULL RzArchPlatformTarget *t, RZ_NON
  * \param t reference to RzArchPlatformTarget
  * \param arch reference to the selected architecture (value of `asm.arch`
  * \param platform reference to the selected platform (value of `asm.platform`)
- * \param dir_prefix reference to the directory prefix or the value of dir.prefix
+ * \param platforms_dir reference to the directory containing platforms files
```
Comment (XVilka) [inline librz/analysis/arch_platform.c]: `platform files` maybe?

File: librz/analysis/arch_profile.c [hunk 2]
```diff
@@ -203,30 +203,28 @@ static bool is_cpu_valid(char *cpu_dir, const char *cpu) {
  * \param t reference to RzArchTarget
  * \param cpu reference to the selected CPU (value of `asm.cpu`)
  * \param arch reference to the selected architecture (value of `asm.arch`)
- * \param dir_prefix reference to the directory prefix or the value of dir.prefix
+ * \param cpus_dir reference to the directory containing cpus files
```
Comment (XVilka) [inline librz/analysis/arch_profile.c]: `CPU files`?

File: librz/cons/dietline.c [hunk 3]
```diff
@@ -465,58 +465,51 @@ RZ_API void rz_line_hist_free(void) {
 }
 
 /* load history from file. TODO: if file == NULL load from ~/.<prg>.history or so */
```
Comment (XVilka) [inline librz/cons/dietline.c]: Maybe remove that `TODO` mark? It's not that relevant anymore, I think.

File: librz/cons/dietline.c [hunk 4]
```diff
@@ -465,58 +465,51 @@ RZ_API void rz_line_hist_free(void) {
...
-RZ_API int rz_line_hist_load(const char *file) {
+RZ_API int rz_line_hist_load(const char *path) {
```
Comment (XVilka) [inline librz/cons/dietline.c]: Maybe check for NULL as well?

File: librz/cons/dietline.c [hunk 5]
```diff
@@ -465,58 +465,51 @@ RZ_API void rz_line_hist_free(void) {
...
-			if (I.history.data) {
-				for (i = 0; i < I.history.index; i++) {
-					fputs(I.history.data[i], fd);
-					fputs("\n", fd);
-				}
-				fclose(fd);
-				ret = true;
-			} else {
-				fclose(fd);
+	char *p = (char *)rz_str_lastbut(path, RZ_SYS_DIR[0], NULL); // TODO: use fs
```
Comment (XVilka) [inline librz/cons/dietline.c]: What is `fs`? Just remove this, IMHO.

File: librz/core/libs.c [hunk 6]
```diff
@@ -51,51 +51,33 @@ CB(parse, parser)
...
-		char *hpd = rz_str_home(RZ_HOME_PLUGINS);
-		if (hpd) {
-			rz_lib_opendir(core->lib, hpd);
-			free(hpd);
-		}
+		char *hpd = rz_path_home(RZ_PLUGINS);
+		rz_lib_opendir(core->lib, hpd, false);
+		free(hpd);
+
+		// TODO: remove after 0.4.0 is released
```
Comment (XVilka) [inline librz/core/libs.c:72]: Don't forget to open an issue to track this after, please

File: librz/include/rz_userconf.h.in [hunk 7]
```diff
@@ -52,42 +54,39 @@
...
-#define RZ_HOME_PROJECTS    RZ_JOIN_2_PATHS(RZ_HOME_DATADIR, "projects")
-#define RZ_HOME_WWWROOT     RZ_JOIN_2_PATHS(RZ_HOME_DATADIR, "www")
-#define RZ_HOME_SDB_TYPES   RZ_JOIN_2_PATHS(RZ_HOME_DATADIR, "types")
-#define RZ_HOME_SDB_FORMAT  RZ_JOIN_2_PATHS(RZ_HOME_DATADIR, "format")
-#define RZ_HOME_HUD         RZ_JOIN_2_PATHS(RZ_HOME_DATADIR, "hud")
-#define RZ_HOME_BINRC       RZ_JOIN_2_PATHS(RZ_HOME_DATADIR, "rc.d")
-#define RZ_HOME_BIN         RZ_JOIN_3_PATHS(RZ_HOME_DATADIR, "prefix", "bin")
-
 #define RZ_HOME_HISTORY RZ_JOIN_2_PATHS(RZ_HOME_CACHEDIR, "history")
+// NOTE: Temporarly added only for compatibility reasons given that the plugin
```
Comment (XVilka) [inline librz/include/rz_userconf.h.in:82]: Same, reference this in the same issue, please.

File: librz/syscall/syscall.c [hunk 8]
```diff
@@ -202,19 +202,21 @@ RZ_API bool rz_sysreg_load_sdb(RzSysregsDB *sysregdb, const char *path) {
...
-	char *path = sdb_fmt(RZ_JOIN_4_PATHS("%s", RZ_SDB, "reg", "%s-%s-%d.sdb"), dir_prefix,
-		arch, s->cpu, s->bits);
-
-	if (path) {
-		if (!rz_sysreg_load_sdb(s->srdb, path)) {
-			rz_sysregs_db_free(s->srdb);
-			s->srdb = rz_sysregs_db_new();
-			return false;
-		}
+RZ_API bool rz_sysreg_set_arch(RzSyscall *s, const char *arch, const char *regs_dir) {
```
Comment (XVilka) [inline librz/syscall/syscall.c]: Check for NULLs please

File: librz/util/file.c [hunk 9]
```diff
@@ -1255,7 +1255,17 @@ RZ_API RzList *rz_file_globsearch(const char *_globbed_path, int maxdepth) {
...
 
-RZ_API char *rz_file_path_join(const char *s1, const char *s2) {
+/**
+ * \brief Concatenate two paths to create a new one with s1+s2 with the correct path separator
+ *
+ * \param s1 First path
+ * \param s2 Second path, can be NULL
+ * \return Full path
+ */
+RZ_API RZ_OWN char *rz_file_path_join(const char *s1, const char *s2) {
```
Comment (XVilka) [inline librz/util/file.c]: Add `RZ_NONNULL` for `s1` and `RZ_NULLABLE` for `s2`, also `rz_return_val_if_fail(s1, NULL)`

File: librz/util/lib.c [hunk 10]
```diff
@@ -374,7 +373,23 @@ RZ_API int rz_lib_open_ptr(RzLib *lib, const char *file, void *handler, RzLibStr
...
+ * Opens all the files ending with the right library extension (e.g. ".so")
+ * present in the directory pointed by \p path . If \p path was already opened,
+ * it is not opened again unless \p force is set to true.
+ *
+ * \param lib Reference to RzLib
+ * \param path Directory to open
+ * \param force When true, a directory is re-scanned even if it was already opened
+ * \return True when the directory is scanned for libs, false otherwise
+ */
+RZ_API bool rz_lib_opendir(RzLib *lib, const char *path, bool force) {
```
Comment (XVilka) [inline librz/util/lib.c:398]: Check for NULL for `path`

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks very good. Please address a few minor issues.

File: librz/core/libs.c (hunk 6 above)
Comment (ret2libc) [reply librz/core/libs.c:72]: https://github.com/rizinorg/rizin/issues/2046

File: librz/include/rz_userconf.h.in (hunk 7 above)
Comment (ret2libc) [reply librz/include/rz_userconf.h.in:82]: https://github.com/rizinorg/rizin/issues/2046

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
```
Run sudo update-alternatives --install /usr/bin/clang-format clang-format /usr/bin/clang-format-13 130
update-alternatives: using /usr/bin/clang-format-13 to provide /usr/bin/clang-format (clang-format) in auto mode
Ubuntu clang-format version 13.0.1-++20211124043118+19b8368225dc-1~exp1~20211124043708.23
librz/util/path.c:76:3: error: code should be clang-formatted [-Wclang-format-violations]
 * 
  ^
librz/cons/dietline.c:469:3: error: code should be clang-formatted [-Wclang-format-violations]
 * 
  ^
librz/cons/dietline.c:495:3: error: code should be clang-formatted [-Wclang-format-violations]
 * 
  ^
```

**Review by thestr4ng3r — APPROVED** [review]

**Review by wargio — APPROVED** [review]

**Comment by XVilka** [comment]
Resolve conflicts please and merge - it passed all CI jobs.

**Pull request merged**